### PR TITLE
Add implementation for Azure Scheduler

### PR DESCRIPTION
**Pull request by milismsft**

### CODE REVIEW

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java [hunk 1]
```diff
@@ -93,33 +98,217 @@
          */
         interface WithStartTime {
             /**
+             * A job which will start right away.
```
Comment (martinsawicki) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: to be consistent with starting the description with a verb: "Specifies that the job should be started right away."

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java [hunk 2]
```diff
@@ -93,33 +98,217 @@
...
         /**
          * The stage of the job definition for an Azure Scheduler allowing to specify the recurrence and action details.
          */
-        interface WithRecurrenceOrAction {
+        interface WithRecurrenceOrAction extends WithAction {
             /**
-             * Specifies the recurrence  of the job.
+             * The recurrence object specifies recurrence rules for the job and the recurrence the job will execute with.
+             * <p>
+             * The recurrence object supports the elements frequency, interval, endTime, count, and schedule. If recurrence
```
Comment (martinsawicki) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: we don't want to use code-like sounding wording here. Better use just plain language:
"The recurrent object allows the frequency, the end time, the number of occurrences and the schedule to be specified. If a recurrence is defined, at least the frequency should be specified, the rest is optional."

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java [hunk 3]
```diff
@@ -93,33 +98,217 @@
...
              *
-             * @param jobRecurrence the start time
+             * @param jobRecurrence the reccurence object
              * @return the next stage of the definition
              */
             WithAction withRecurrence(JobRecurrence jobRecurrence);
 
             /**
-             * Specifies the action details of the job.
+             * Specifies the interval for the frequency that determines how often the job will run.
```
Comment (martinsawicki) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: This description seems to use the Swagger's confusing naming, which even I have a hard time following. It'd be good to explain this in simpler terms. How about something like: 
```
Specifies the number of time units between consecutive occurrences. The exact time unit to use is specified at the next stage.
<p>
For example, if 3 is specified here and "week" is selected at the next stage, the job will recur every 3 weeks.
<p>
Different time units support different maximum numbers of occurrences. Those maximum combinations are: 18 months, 78 weeks, 548 days, 1000 hours or minutes.
```

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java [hunk 4]
```diff
@@ -93,33 +98,217 @@
...
+             */
+            WithAdvanceDaysOfMonthSchedule runningMonthly();
+        }
+
+        /**
+         * The stage of the job definition for an Azure Scheduler allowing to specify the frequency unit.
+         */
+        interface WithRecurrenceFrequencyAndSchedule {
+            /**
+             * Specifies the recurrence of the job to be every minute.
```
Comment (martinsawicki) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: I don;t think this text is correct. Shouldn't it be more like: 
"Selects minutes as the time unit for the number of occurrences."

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java [hunk 5]
```diff
@@ -93,33 +98,217 @@
...
+             *
+             * @return the next stage of the definition
+             */
+            WithEndTime minutes();
+
+            /**
+             * Specifies the recurrence of the job to be every hour.
```
Comment (martinsawicki) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: similarly:
"Selects hours as the time unit for the number of occurrences."

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java [hunk 6]
```diff
@@ -93,33 +98,217 @@
...
+             *
+             * @return the next stage of the definition
+             */
+            WithEndTime hours();
+
+            /**
+             * Specifies the recurrence of the job to be every day.
```
Comment (martinsawicki) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: similarly:
"Selects days as the time unit for the number of occurrences."

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java [hunk 7]
```diff
@@ -93,33 +98,217 @@
...
+             *
+             * @return the next stage of the definition
+             */
+            WithAdvanceMinutesAndHoursSchedule days();
+
+            /**
+             * Specifies the recurrence of the job to be every week.
```
Comment (martinsawicki) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: similarly:
"Selects weeks as the time unit for the number of occurrences."

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java [hunk 8]
```diff
@@ -93,33 +98,217 @@
...
+             *
+             * @return the next stage of the definition
+             */
+            WithAdvanceDaysOfWeekSchedule weeks();
+
+            /**
+             * Specifies the recurrence  of the job to be every month.
```
Comment (martinsawicki) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: similarly:
"Selects months as the time unit for the number of occurrences."

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java [hunk 9]
```diff
@@ -93,33 +98,217 @@
...
+        interface WithEndTime extends WithAction {
+            /**
+             * Specifies the job to run indefinitely.
+             *
+             * @return the next stage of the definition
+             */
+            WithAction endingNever();
+
+            /**
+             * Specifies the job to complete at a given time.
```
Comment (martinsawicki) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: "complete" I think means something different than what is intended here. This is about when the recurrence will end, not the job.  So how about:
"Specifies the time by which no more occurrences of this job will be started."

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java [hunk 10]
```diff
@@ -93,33 +98,217 @@
...
+             *
+             * @param endTime number of times the job will run
+             * @return the next stage of the definition
+             */
+            WithAction endingBy(DateTime endTime);
+
+            /**
+             * Specifies the job to complete after the specified number of runs.
```
Comment (martinsawicki) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: similarly:
"Specifies the number of occurrences after which the job will no longer be re-started."

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/JobScheduleMonthlyWeekDay.java [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+ * Licensed under the MIT License. See License.txt in the project root for
+ * license information.
+ */
+package com.microsoft.azure.management.scheduler;
+
+import com.microsoft.azure.management.apigeneration.Beta;
+import com.microsoft.azure.management.apigeneration.Fluent;
+
+/**
+ * The day of the week the job will run each month on weekly bases.
```
Comment (martinsawicki) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/JobScheduleMonthlyWeekDay.java]: maybe a little more accurate/clearer:
"Days of the week relative to the beginning of the month for scheduling a job."

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/JobScheduleMonthlyWeekDay.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+    FOURTH_SUNDAY(4, JobScheduleDay.SUNDAY),
+    LAST_MONDAY(-1, JobScheduleDay.MONDAY),
+    LAST_TUESDAY(-1, JobScheduleDay.TUESDAY),
+    LAST_WEDNESDAY(-1, JobScheduleDay.WEDNESDAY),
+    LAST_THURSDAY(-1, JobScheduleDay.THURSDAY),
+    LAST_FRIDAY(-1, JobScheduleDay.FRIDAY),
+    LAST_SATURDAY(-1, JobScheduleDay.SATURDAY),
+    LAST_SUNDAY(-1, JobScheduleDay.SUNDAY);
+
+    /** The actual serialized value for a JobScheduleMonthlyWeekDay instance. */
```
Comment (martinsawicki) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/JobScheduleMonthlyWeekDay.java]: this is a private field, so this should not be a JavaDoc (/**) but a regular code comment (//)

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/implementation/JobImpl.java [hunk 13]
```diff
@@ -147,4 +250,289 @@ public JobImpl withState(JobState state) {
 
         return this;
     }
+
+    @Override
+    public JobImpl minutes() {
+        if (this.inner().properties() == null) {
```
Comment (martinsawicki) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/implementation/JobImpl.java]: considering that these 3 or so lines of code are repeated several times across the next few withers, wouldn't it be cleaner to just have a private `ensureProperties()` function which returne the inner properties object, ensure one is present?
And then simply refer to it in the specifici wither impls as:
`ensureProperties().withRecurrence(...)`
etc.

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/implementation/JobImpl.java [hunk 14]
```diff
@@ -147,4 +250,289 @@ public JobImpl withState(JobState state) {
...
+
+        return this;
+    }
+
+    private void resetJobScheduleDaysList() {
+        // Only one days list can be specified, other scheduler lists need to be set to null
+        if (this.inner().properties() != null &&
+            inner().properties().recurrence() != null &&
+            inner().properties().recurrence().schedule() != null) {
+            this.inner().properties()
```
Comment (martinsawicki) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/implementation/JobImpl.java]: the repetition/verbosity seems redundant. Why not simply:

```java
this.innerProperties().recurrence().schedule()
    .withWeekDays(null)
    .withMonthDays(null)
    .withMonthlyOccurrences(null);
```

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/implementation/JobImpl.java [hunk 15]
```diff
@@ -147,4 +250,289 @@ public JobImpl withState(JobState state) {
...
+                .recurrence()
+                .schedule()
+                .withMonthlyOccurrences(null);
+        }
+    }
+
+    @Override
+    public JobImpl onTheseDays(DayOfWeek... days) {
+        if (this.inner().properties() == null) {
```
Comment (martinsawicki) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/implementation/JobImpl.java]: along lines similar to a previous comments, it seems a bunch of repeated code could be put inside a private helper method `ensureSchedule()` across the next several withers.

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/implementation/JobImpl.java [hunk 16]
```diff
@@ -147,4 +250,289 @@ public JobImpl withState(JobState state) {
...
+        }
+        for (int hour : hours) {
+            this.inner().properties().recurrence().schedule().hours().add(hour);
+        }
+
+        return this;
+    }
+
+    @Override
+    public JobImpl endingNever() {
```
Comment (martinsawicki) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/implementation/JobImpl.java]: along similar lines to before, it seems a helper `ensureRecurrence()` would help save a whole bunch of repeated code in the next few withers.

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java (hunk 1 above)
Comment (milismsft) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: fixed

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java (hunk 2 above)
Comment (milismsft) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: Fixed (though it was text copied and paste from the Scheduler docs...)

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java (hunk 4 above)
Comment (milismsft) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: fixed

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java (hunk 5 above)
Comment (milismsft) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: fixed

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java (hunk 6 above)
Comment (milismsft) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: fixed

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java (hunk 7 above)
Comment (milismsft) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: fixed

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java (hunk 8 above)
Comment (milismsft) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: fixed

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java (hunk 9 above)
Comment (milismsft) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: not quite; the job can still be run manually or if the recurrence is updated and the state is not "completed". I'll try to use a better wording.

Comment (milismsft) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: how about: "Specifies the time after which the job will no longer run recurrently."

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java (hunk 10 above)
Comment (milismsft) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: ..and "Specifies the specified number of runs after which the job will no longer run recurrently."

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/JobScheduleMonthlyWeekDay.java (hunk 11 above)
Comment (milismsft) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/JobScheduleMonthlyWeekDay.java]: fixed

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/JobScheduleMonthlyWeekDay.java (hunk 12 above)
Comment (milismsft) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/JobScheduleMonthlyWeekDay.java]: checkstyle check fails if the enum is not documented that way...

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/implementation/JobImpl.java (hunk 13 above)
Comment (milismsft) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/implementation/JobImpl.java]: makes sense.

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java (hunk 9 above)
Comment (martinsawicki) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: or maybe simpler language to say the same thing:
"Specifies the time after which the job will no longer be started automatically."
?

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java (hunk 10 above)
Comment (martinsawicki) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: so maybe:
"Specifies the number of runs after which the job will no longer be started automatically."
?

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/implementation/JobImpl.java (hunk 15 above)
Comment (milismsft) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/implementation/JobImpl.java]: fixed

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/implementation/JobImpl.java (hunk 16 above)
Comment (milismsft) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/implementation/JobImpl.java]: fixed

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/implementation/JobImpl.java (hunk 14 above)
Comment (milismsft) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/implementation/JobImpl.java]: fixed

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java (hunk 3 above)
Comment (milismsft) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: fixed

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java (hunk 9 above)
Comment (milismsft) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: sounds good.

File: azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java (hunk 10 above)
Comment (milismsft) [inline azure-mgmt-scheduler/src/main/java/com/microsoft/azure/management/scheduler/Job.java]: ok
